runtime/commonMain/src/kotlinx/serialization/internal/CollectionSerializers.kt

```diff
@@ -140,27 +128,74 @@ sealed class MapLikeSerializer<TKey, TVal, TCollection, TBuilder: MutableMap<TKe
     }
 }
 
+public abstract class PrimitiveArraySerializer<Element, Array, Builder : PrimitiveArraySerializer<Element, Array, Builder>.Builder>
+internal constructor(
+    primitiveSerializer: KSerializer<Element>,
+    primitiveDescriptor: PrimitiveDescriptor
+) : ListLikeSerializer<Element, Array, Builder>(primitiveSerializer) {
+    final override val descriptor = PrimitiveArrayDescriptor(primitiveDescriptor)
+
+    abstract inner class Builder internal constructor() {
+        abstract val position: Int
+        abstract fun ensureCapacity(requiredCapacity: Int = position * 2 + 1)
+        abstract fun build(): Array
+    }
+
+    final override fun Builder.builderSize() = position
+    final override fun Builder.toResult(): Array = build()
+    final override fun Builder.checkCapacity(size: Int) = ensureCapacity(size)
+
+    final override fun Array.collectionIterator(): Iterator<Element> =
+        error("This method lead to boxing and must not be used, use writeContents instead")
+
+    final override fun Builder.insert(index: Int, element: Element) =
+        error("This method lead to boxing and must not be used, use Builder.append instead")
+
+    protected abstract override fun readItem(
+        decoder: CompositeDecoder,
+        index: Int,
+        builder: Builder,
+        checkIndex: Boolean
+    )
+
+    protected abstract fun writeContents(encoder: CompositeEncoder, content: Array, size: Int)
+
+    final override fun serialize(encoder: Encoder, obj: Array) {
+        val size = obj.collectionSize()
+        @Suppress("NAME_SHADOWING")
+        val encoder = encoder.beginCollection(descriptor, size, *typeParams)
+        writeContents(encoder, obj, size)
+        encoder.endStructure(descriptor)
+    }
+}
+
 // todo: can be more efficient when array size is know in advance, this one always uses temporary ArrayList as builder
-class ReferenceArraySerializer<T: Any, E: T?>(private val kClass: KClass<T>, eSerializer: KSerializer<E>):
-        ListLikeSerializer<E, Array<E>, ArrayList<E>>(eSerializer) {
+class ReferenceArraySerializer<ElementKlass : Any, Element : ElementKlass?>(
```

Comment:
Since it's effectively compiled into the client's code, it is incorrect to insert internal here.

It is also not unused: https://github.com/kotlin/kotlinx.serialization/blob/45aa7c7ab97f25f205408b46a7069886408bce6b/runtime/jvmMain/src/kotlinx/serialization/JvmResolving.kt#L58

Reply:
Thanks.
`@InternalSerializationApi`?